# PANDAS INTRODUCTION

Pandas adalah tools analisis dan manipulasi data open source yang cepat, powerful, fleksibel, dan mudah digunakan, dibangun di atas bahasa pemrograman Python.

Pandas adalah pustaka Python yang menyediakan sarana ekstensif untuk analisis data. Data scientist sering kali bekerja dengan data yang disimpan dalam format tabel seperti .csv, .tsv, atau .xlsx. Pandas membuatnya sangat nyaman untuk load, memproses, dan menganalisis data tabel tersebut menggunakan kueri seperti SQL. Dalam hubungannya dengan Matplotlib dan Seaborn, Pandas memberikan berbagai peluang untuk analisis visual data tabel.

Struktur data utama di Pandas diimplementasikan dengan Series dan DataFrame. DataFrames sangat bagus untuk merepresentasikan data nyata: baris sesuai dengan instance (objek, observasi, dll.), dan kolom sesuai dengan features untuk setiap instance.

In [2]:
import pandas as pd 
import numpy as np

In [3]:
df = pd.read_csv('nbaallelo.csv')
df

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
0                1            0     TRH    Huskies  ...  40.294830     NYK   
1                1            0     NYK     Knicks  ...  41.705170     TRH   
2                1            0     CHS      Stags  ...  42.012257     NYK   
3                2            0     NYK     Knicks  ...  40.692783     CHS   
4                1            0     DTF    Falcons  ...  38.864048     WSC   
...            ...          ...     ...        ...  ...        ...     ...   
126309         100            1     CLE  Cavaliers  ...  60.309792     GSW   
126310         102            1     GSW   Warriors  ...  68.013329     CLE   
126311         101            1     CLE  Cavaliers  ...  60.010067     GSW   
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
0          Knicks       68  1300.0000  1306.7233              H            L   
1         Huskies       66  1300.0000  1293.2767              A            W   
2          Knicks       47  1306.7233  1297.0712              H            W   
3           Stags       63  1300.0000  1309.6521              A            L   
4        Capitols       50  1300.0000  1320.3811              H            L   
...           ...      ...        ...        ...            ...          ...   
126309   Warriors      103  1790.9591  1809.9791              H            L   
126310  Cavaliers       91  1704.3949  1700.7391              H            W   
126311   Warriors      104  1809.9791  1813.6349              A            L   
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
0       0.640065   NaN  
1       0.359935   NaN  
2       0.631101   NaN  
3       0.368899   NaN  
4       0.640065   NaN  
...          ...   ...  
126309  0.546572   NaN  
126310  0.765565   NaN  
126311  0.234435   NaN  
126312  0.481450   NaN  
126313  0.518550   NaN  

[126314 rows x 23 columns]

In [4]:
len(df)

126314

In [5]:
df.shape

(126314, 23)

Kalian menggunakan fungsi len() bawaan Python untuk melihat jumlah baris. Kalian juga menggunakan atribut .shape DataFrame untuk melihat dimensinya. Hasilnya adalah tuple yang berisi jumlah baris dan kolom.

Sekarang kalian tahu bahwa ada 126.314 baris dan 23 kolom dalam kumpulan data kalian.

Tapi bagaimana kalian bisa yakin bahwa kumpulan data benar-benar berisi basketball stats? kalian dapat melihat lima baris pertama dengan .head():

In [6]:
df.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In [7]:
pd.set_option("display.max.columns", None)

In [8]:
pd.set_option("display.precision", 2)

Untuk memverifikasi bahwa kalian telah berhasil mengubah opsi, kalian dapat menjalankan .head() lagi, atau kalian dapat menampilkan lima baris terakhir dengan .tail() sebagai gantinya:

In [9]:
df.tail()

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts    elo_i    elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.41  1704.39   
126310         102            1     GSW   Warriors  104  1809.98  1813.63   
126311         101            1     CLE  Cavaliers   91  1704.39  1700.74   
126312         102            1     CLE  Cavaliers   97  1700.74  1692.09   
126313         103            1     GSW   Warriors  105  1813.63  1822.29   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309      60.31    GSW   Warriors      103    1790.96    1809.98   
126310      68.01    CLE  Cavaliers       91    1704.39    1700.74   
126311      60.01    GSW   Warriors      104    1809.98    1813.63   
126312      59.29    GSW   Warriors      105    1813.63    1822.29   
126313      68.52    CLE  Cavaliers       97    1700.74    1692.09   

       game_location game_result  forecast notes  
126309             H           L      0.55   NaN  
126310             H           W      0.77   NaN  
126311             A           L      0.23   NaN  
126312             H           L      0.48   NaN  
126313             A           W      0.52   NaN

Kalian dapat menemukan beberapa posibilitas.head() dan .tail() lebih lanjut dengan latihan kecil. Dapatkah kalian mencetak tiga baris terakhir DataFrame kalian?

In [10]:
df.tail(3)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts    elo_i    elo_n  \
126311         101            1     CLE  Cavaliers   91  1704.39  1700.74   
126312         102            1     CLE  Cavaliers   97  1700.74  1692.09   
126313         103            1     GSW   Warriors  105  1813.63  1822.29   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126311      60.01    GSW   Warriors      104    1809.98    1813.63   
126312      59.29    GSW   Warriors      105    1813.63    1822.29   
126313      68.52    CLE  Cavaliers       97    1700.74    1692.09   

       game_location game_result  forecast notes  
126311             A           L      0.23   NaN  
126312             H           L      0.48   NaN  
126313             A           W      0.52   NaN

Dalam file data nbaallelo.csv terdapat variabel berikut:

## Getting to Know Your Data

Kalian telah mengimpor file CSV dengan pustaka Pandas Python dan melihat konten kumpulan data kalian untuk pertama kalinya. Sejauh ini, kalian hanya melihat ukuran kumpulan data kalian dan beberapa baris pertama dan terakhirnya. Selanjutnya, kalian akan mempelajari cara memeriksa data kalian secara lebih sistematis.

## Displaying Data Types
Langkah pertama untuk mengetahui data kalian adalah menjelajahi berbagai tipe data yang dikandungnya. Meskipun kalian bisa memasukkan apa pun ke dalam list, kolom DataFrame berisi nilai dari tipe data tertentu. Saat kalian membandingkan struktur data Pandas dan Python, kalian akan melihat bahwa behavior ini membuat Pandas jauh lebih cepat!

Kalian bisa menampilkan semua kolom dan tipe datanya dengan .info():

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

Kalian akan melihat daftar semua kolom dalam kumpulan data kalian dan jenis data yang ada di setiap kolom. Di sini, kalian bisa melihat tipe data int64, float64, dan object. Pandas menggunakan perpustakaan NumPy untuk bekerja dengan tipe ini. Nanti, kalian akan menemukan tipe data kategorikal yang lebih kompleks, yang diimplementasikan oleh pustaka Pandas.

## Showing Basics Statistics
Setelah kalian melihat jenis data apa yang ada dalam kumpulan data kalian , sekarang saatnya untuk mendapatkan overview tentang nilai yang terkandung di setiap kolom. Kalian dapat melakukan ini dengan .describe():

In [12]:
df.describe()

gameorder   _iscopy    year_id  seasongame  is_playoffs        pts  \
count  126314.00  126314.0  126314.00   126314.00    126314.00  126314.00   
mean    31579.00       0.5    1988.20       43.53         0.06     102.73   
std     18231.93       0.5      17.58       25.38         0.24      14.81   
min         1.00       0.0    1947.00        1.00         0.00       0.00   
25%     15790.00       0.0    1975.00       22.00         0.00      93.00   
50%     31579.00       0.5    1990.00       43.00         0.00     103.00   
75%     47368.00       1.0    2003.00       65.00         0.00     112.00   
max     63157.00       1.0    2015.00      108.00         1.00     186.00   

           elo_i      elo_n  win_equiv    opp_pts  opp_elo_i  opp_elo_n  \
count  126314.00  126314.00  126314.00  126314.00  126314.00  126314.00   
mean     1495.24    1495.24      41.71     102.73    1495.24    1495.24   
std       112.14     112.46      10.63      14.81     112.14     112.46   
min      1091.64    1085.77      10.15       0.00    1091.64    1085.77   
25%      1417.24    1416.99      34.10      93.00    1417.24    1416.99   
50%      1500.95    1500.95      42.11     103.00    1500.95    1500.95   
75%      1576.06    1576.29      49.64     112.00    1576.06    1576.29   
max      1853.10    1853.10      71.11     186.00    1853.10    1853.10   

        forecast  
count  126314.00  
mean        0.50  
std         0.22  
min         0.02  
25%         0.33  
50%         0.50  
75%         0.67  
max         0.98

## Exploring Your Dataset
Exploratory data analysis dapat membantu kalian menjawab pertanyaan tentang kumpulan data kalian. Misalnya, kalian dapat memeriksa seberapa sering nilai tertentu muncul dalam kolom:

In [13]:
df["team_id"].value_counts()

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
INJ      60
PIT      60
DTF      60
TRH      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [14]:
df["fran_id"].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

Tampaknya tim bernama "Lakers" memainkan 6024 pertandingan, tetapi hanya 5078 yang dimainkan oleh Los Angeles Lakers. Cari tahu siapa tim "Lakers" lainnya:

In [15]:
df.loc[df["fran_id"] == "Lakers", "team_id"].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

In [16]:
df.loc[df["team_id"] == "MNL", "date_game"].min()

'1/1/1949'

In [17]:
df.loc[df["team_id"] == "MNL", "date_game"].max()

'4/9/1959'

In [18]:
df.loc[df["team_id"] == "MNL", "date_game"].agg(("min", "max"))

min    1/1/1949
max    4/9/1959
Name: date_game, dtype: object

Sepertinya Minneapolis Lakers dimainkan main tahun 1949 dan 1959. Itu menjelaskan mengapa kalian mungkin tidak mengenali tim ini!

Kalian juga telah mengetahui mengapa tim Boston Celtics "BOS" memainkan game paling banyak dalam kumpulan data. Mari kita analisis sedikit tentang sejarah mereka. Cari tahu berapa banyak poin yang telah dicetak Boston Celtics selama semua pertandingan yang terdapat dalam dataset ini.

In [19]:
df.loc[df["team_id"] == "BOS", "pts"].sum()

626484

Boston Celtics mencetak total 626.484 poin.

## Getting to Know Pandas’ Data Structures
Meskipun DataFrame menyediakan fungsi yang bisa terasa cukup intuitif, konsep dasarnya agak lebih rumit untuk dipahami. Oleh karena itu, kalian akan menyisihkan NBA DataFrame yang luas dan membuat beberapa objek Pandas yang lebih kecil dari awal.

## Understanding Series Objects
Struktur data Python yang paling dasar adalah list, yang juga merupakan titik awal yang baik untuk mengenal objek pandas.Series. Buat objek Series baru menggunakan list:

In [20]:
revenues = pd.Series([5555, 7000, 1980])
revenues

0    5555
1    7000
2    1980
dtype: int64

Kalian menggunakan list [5555, 7000, 1980] untuk membuat objek Series yang disebut revenues. Objek Series mempunyai dua komponen:

* A sequence of values
* A sequence of identifiers, which is the index

Kalian dapat mengakses komponen ini masing-masing dengan .values dan.index:

revenues.values mengembalikan nilai dalam Seri, sedangkan revenues.index mengembalikan indeks posisi.

In [21]:
revenues.values

array([5555, 7000, 1980], dtype=int64)

In [22]:
revenues.index

RangeIndex(start=0, stop=3, step=1)

Meskipun Pandas dibuat di atas NumPy, perbedaan yang signifikan terletak pada pengindeksannya. Sama seperti array NumPy, Pandas Series juga memiliki indeks integer yang ditentukan secara implisit. Indeks implisit ini menunjukkan posisi elemen dalam Series.

Namun, Series juga dapat memiliki jenis indeks yang berubah-ubah. Kalian dapat menganggap indeks eksplisit ini sebagai label untuk baris tertentu:

In [23]:
city_revenues = pd.Series(
    [4200, 8000, 6500],
    index=["Amsterdam", "Toronto", "Tokyo"]
)

city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

Di sini, indeks adalah daftar nama kota yang diwakili oleh string. Kalian mungkin telah memperhatikan bahwa dictionaries Python menggunakan indeks string juga, dan ini adalah analogi yang berguna untuk diingat! Kalian dapat menggunakan blok kode di atas untuk membedakan antara dua jenis Series:

* revenues: This Series behaves like a Python list because it only has a positional index.
* city_revenues: This Series acts like a Python dictionary because it features both a positional and a label index.

Berikut cara membuat Series dengan indeks label dari dictionary Python:

In [24]:
city_employee_count = pd.Series({"Amsterdam": 5, "Tokyo": 8})
city_employee_count

Amsterdam    5
Tokyo        8
dtype: int64

Dictionary keys menjadi indeks, dan nilai dictionary adalah nilai Series.

Sama seperti dictionaries, Series juga mendukung .keys() dan keyword in:

In [25]:
city_employee_count.keys()

Index(['Amsterdam', 'Tokyo'], dtype='object')

In [26]:
"Tokyo" in city_employee_count

True

In [27]:
"New York" in city_employee_count

False

Kalian dapat menggunakan metode ini untuk menjawab pertanyaan tentang kumpulan data kalian dengan cepat.

## Understanding DataFrame Objects
Meskipun Series adalah struktur data yang cukup kuat, ia memiliki keterbatasan. Misalnya, kalian hanya dapat menyimpan satu atribut per kunci. Seperti yang kalian lihat dengan kumpulan data nba, yang menampilkan 23 kolom, pustaka Pandas Python memiliki lebih banyak hal untuk ditawarkan dengan DataFrame-nya. Struktur data ini adalah sequence objek kalian yang memiliki indeks yang sama.

Jika kalian telah mengikuti contoh Series diatas, kalian seharusnya sudah memiliki dua objek Series dengan cities sebagai kuncinya:

* city_revenues
* city_employee_count

Kalian dapat menggabungkan objek-objek ini ke dalam DataFrame dengan memberikan dictionary di konstruktor. Dictionary keys akan menjadi nama kolom, dan nilainya akan berisi objek Series:

In [28]:
city_data = pd.DataFrame({
    "revenue" : city_revenues,
    "employee_count": city_employee_count
})

In [29]:
city_data

revenue  employee_count
Amsterdam     4200             5.0
Tokyo         6500             8.0
Toronto       8000             NaN

Perhatikan bagaimana Pandas mengganti nilai employee_count yang hilang untuk Toronto dengan NaN.

Indeks DataFrame baru adalah gabungan dari dua indeks Series:

In [30]:
city_data.index

Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object')

Sama seperti Series, DataFrame juga menyimpan nilainya dalam array NumPy:

In [31]:
city_data.values

array([[4.2e+03, 5.0e+00],
       [6.5e+03, 8.0e+00],
       [8.0e+03,     nan]])

Kalian juga dapat merujuk ke 2 dimensi DataFrame sebagai axes:

In [32]:
city_data.axes

[Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object'),
 Index(['revenue', 'employee_count'], dtype='object')]

In [33]:
city_data.axes[0]

Index(['Amsterdam', 'Tokyo', 'Toronto'], dtype='object')

Axis bertanda 0 adalah indeks baris, dan axis bertanda 1 adalah indeks kolom. Terminologi ini penting untuk diketahui karena kalian akan menemukan beberapa metode DataFrame yang menerima parameter axis.

DataFrame juga merupakan struktur data seperti dictionary, sehingga juga mendukung .keys() dan keyword in. Namun, untuk DataFrame, hal ini tidak terkait dengan indeks, tetapi dengan kolom:

In [34]:
city_data.keys()

Index(['revenue', 'employee_count'], dtype='object')

In [35]:
"Amsterdam" in city_data

False

In [36]:
"revenue" in city_data

True

Kalian dapat melihat konsep ini diimplementasikan dengan kumpulan data NBA yang lebih besar. Apakah dataset tersebut berisi kolom yang disebut "points", atau apakah disebut "pts"?

Karena kalian tidak menentukan kolom indeks saat kalian membaca di file CSV, Pandas telah menetapkan RangeIndex ke DataFrame:

In [37]:
df.index

RangeIndex(start=0, stop=126314, step=1)

df, seperti semua objek DataFrame, memiliki dua axes:

In [38]:
df.axes

[RangeIndex(start=0, stop=126314, step=1),
 Index(['gameorder', 'game_id', 'lg_id', '_iscopy', 'year_id', 'date_game',
        'seasongame', 'is_playoffs', 'team_id', 'fran_id', 'pts', 'elo_i',
        'elo_n', 'win_equiv', 'opp_id', 'opp_fran', 'opp_pts', 'opp_elo_i',
        'opp_elo_n', 'game_location', 'game_result', 'forecast', 'notes'],
       dtype='object')]

Kalian dapat memeriksa keberadaan kolom dengan .keys():

In [39]:
"points" in df.keys()

False

In [40]:
"pts" in df.keys()

True

Kolom tersebut disebut dengan "pts", bukan "points".

## Accessing Series Elements
Pada bagian di atas, kalian telah membuat Series Pandas berdasarkan list Python dan membandingkan kedua struktur datanya. Kalian telah melihat kemiripan objek Series dengan list dan dictionaries dalam beberapa cara. Kesamaan lebih lanjut adalah kalian dapat menggunakan operator pengindeksan ([]) untuk Series juga.

Kalian juga akan mempelajari cara menggunakan dua metode akses khusus Pandas:

* .loc
* .iloc
Kalian akan melihat bahwa metode akses data ini bisa jauh lebih mudah dibaca daripada operator pengindeksan.

**Using the Indexing Operator**

Ingatlah bahwa Series memiliki dua indeks:

* A positional or implicit index, which is always a RangeIndex
* A label or explicit index, which can contain any hashable objects

Selanjutnya, kunjungi kembali objek city_revenues:

In [41]:
city_revenues

Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

Kalian dapat dengan mudah mengakses nilai dalam Series dengan indeks label dan posisi:

In [43]:
city_revenues["Toronto"]

8000

In [44]:
city_revenues[-1]

6500

In [45]:
city_revenues[1:]

Toronto    8000
Tokyo      6500
dtype: int64

In [46]:
city_revenues["Toronto":]

Toronto    8000
Tokyo      6500
dtype: int64

**Using .loc and .iloc**

Operator pengindeksan ([]) memang mudah digunakan, tetapi ada peringatan. Bagaimana jika labelnya juga angka? Katakanlah kalian harus bekerja dengan objek Series seperti ini:

In [48]:
colors = pd.Series(
    ["red", "purple", "blue", "green", "yellow"],
    index=[1, 2, 3, 4, 5]
)

colors

1       red
2    purple
3      blue
4     green
5    yellow
dtype: object

Apakah output dari colors[1]? Untuk indeks posisi, colors[1] adalah "ungu". Namun, jika kalian menggunakan indeks label, maka colors[1] mengacu pada "merah".

Untuk menghindari kebingungan, pustaka Pandas Python menyediakan dua metode akses data:

* .loc refers to the label index.
* .iloc refers to the positional index.

Metode akses data ini jauh lebih mudah dibaca:

In [49]:
colors.loc[1]

'red'

In [50]:
colors.iloc[1]

'purple'